Sharp carreau lethe example (#487)
lethe-cfd/lethe: doc/source/examples/resolved_cfd-dem/sphere-carreau-with-sharp-interface/sphere-carreau-with-sharp-interface.rst

.. image:: images/sharp_carreau_case.png
    :alt: Simulation schematic
    :align: center

Parameter file
-----------------------

Mesh
~~~~~

The mesh is defined using the following subsection.

.. code-block:: text

	subsection mesh
	  set type                 = dealii
	  set grid type            = subdivided_hyper_rectangle
	  set grid arguments       = 2,1,1 : -18,-15,-15 : 42,15,15 : true
	  set initial refinement   = 4
	end
	
The dimensions of the used domain are (60x30x30), and the ``subdivided_hyper_rectangle`` is initially divided in (2x1x1) cells, the cells are therefore cubic and of initial size (30x30x30). Using an ``initial refinement`` of 4, the initial size of the cubic cells is 30/2^4 = 1.875. Since the particle size is small in regards to the mesh size, a refinement zone is generated around the particle to better capture it (see :doc:`../../../parameters/cfd/box_refinement` for more details).

.. code-block:: text

	subsection  box refinement
	  set initial refinement   = 3
	  subsection mesh
	      set type                 = dealii
	      set grid type            = subdivided_hyper_rectangle
	      set grid arguments       = 1,1,1: -2,-2,-2 : 6,2,2 : true

> 	      set initial refinement   = 0

Author: Very good point :) the initial refinement of the mesh itself is zero, it should always be as small as possible since when using an box refinement, only the mesh boundaries matter. The time taken to refine the mesh inside the boundaries of the box is O(n_cells_of_mesh*n_cells_in_box_mesh), so it's just much faster to generate the same box when using fewer box mesh cells. Let me know if this still isn't clear. The ``initial refinement = 3`` sets how many times the cells inside the boundaries of the hyper_rectangle are refined.

> 	      set initial refinement   = 0

Author: I also had a hard time understanding at first!

	  end
	end

Boundary conditions
~~~~~~~~~~~~~~~~~~~~
We define the boundary conditions in order to have an inlet velocity of 1 m/s on the left, ``slip`` boundary conditions parallel to the flow direction, and an outlet on the right of the domain (unspecified condition at ``id = 1``).

.. code-block:: text

	subsection boundary conditions
	  set number                  = 5
	  subsection bc 0
		set id 		= 0
		set type    = function
		subsection u
		  set Function expression = 1
		end
		subsection v
		  set Function expression = 0
		end
		subsection w
		  set Function expression = 0
		end
	  end

> 	  subsection bc 1

Author: you are right, I will change that!

> 	  subsection bc 1

Author: Since the number of boundary conditions is 5, I have to list them from 0 to 4. Also, since using a dealii mesh, I have to use the pregiven boundary indices. Therefore, I can't really make them match as it is.

> 	  subsection bc 1

Reviewer: That's all right. I think I would explain why you are doing this though. like just add like you just mentioned. Since we are using a deal.II mesh, we don't have control over the ids :)

.. image:: images/velocity.png
	:align: center

.. image:: images/pressure.png
	:align: center

We can also see the viscosity profile throughout the domain, that is a function of the shear rate magnitude profile. Close to the particle, the shear rate is high which decreases the viscosity. 

.. image:: images/viscosity.png
	:align: center

.. image:: images/shear_rate.png
	:align: center

We can notice that the viscosity rapidely reaches a plateau at :math:`\eta=0.063`. Given the parameters in the **Physical properties** section, the viscosity behavior should be given  by:

.. image:: images/carreau.png
	:align: center

We get the following torques and forces applied on the particle for each of the mesh refinements.

.. code-block:: text

    particle_ID    T_x       T_y       T_z      f_x      f_y       f_z    
	0 	-0.000008  0.000019 -0.000021 0.411248 0.000019  0.000038 
	0 	 0.000000  0.000001 -0.000003 0.415503 0.000001  0.000002 
	0 	-0.000000 -0.000000 -0.000000 0.424717 0.000002 -0.000002 
          
.. note:: 

> 	Because this analysis concerns non-Newtonian flow, there is no known solution for the drag coefficient. For a Newtonian flow at Re = 50, the drag force would be 0.6165. Therefore, the drag force was decreased using a shear-thinning fluid.

Author: It is the f_x, I will add that :)

Possibilities for extension
-----------------------------	
* **High-order methods** : Lethe supports higher order interpolation. This can yield much better results with an equal number of degrees of freedom than traditional second-order (Q1-Q1) methods, especially at higher Reynolds numbers.

> * **Reynolds number** : By changing the inlet velocity, it can be interesting to see the impact of the shear-thinning behavior on the effective drag force.

Reviewer 2: Another extension could be changing the parameters of the fluid? What if we change it to shear-thickening to see the drag force increase?

Author: That's s good point, I will add that. I should also say that the Carreau model is not suitable for shear thickening flows.

We use an *a priori* Reynolds number, since it is not possible, *a priori*, to know the effective viscosity of the flow. For the given parameters, the *a priori* Reynolds number is :math:`50`. 

Initial conditions
~~~~~~~~~~~~~~~~~~~~

This example uses a ramping initial condition that first ramps on the ``n`` parameter, and then on the ``viscosity_0`` parameter. This allows for a smooth transition of non-Newtonian behavior level and of regime.

.. code-block:: text

	subsection initial conditions
	  set type = ramp
	  subsection ramp
		subsection n
		  set initial n 	= 1.0
		  set iterations 	= 2
		  set alpha 		= 0.5
		end
		subsection viscosity
		  set initial viscosity = 1.0
		  set iterations 	= 2
		  set alpha 		= 0.5
		end
	  end
	end
	
The first initial condition simulation solves for ``n=1.0``, ``viscosity_0 = 1.0``, ``viscosity_inf = 0``, ``lambda=10`` and ``a=2``. The subsequent initial simulations are:

* (Second ``n`` iteration) ``n=0.75``, ``viscosity_0 = 1.0``, ``viscosity_inf = 0``, ``lambda=10`` and ``a=2`` ;
* (First ``viscosity`` iteration) ``n=0.5``, ``viscosity_0 = 1.0``, ``viscosity_inf = 0``, ``lambda=10`` and ``a=2`` ;

> * (Second ``viscosity`` iteration) ``n=0.5``, ``viscosity_0 = 0.531702``, ``viscosity_inf = 0``, ``lambda=10`` and ``a=2`` 

Reviewer: I would explain here what is the meaning of the iterations parameters so that people understand the origin of the 2 :)

Author: The equation is in the initial conditions documentation, should it also be added here?

Reviewer: Nah I think it's ok.

* (First ``viscosity`` iteration) ``n=0.5``, ``viscosity_0 = 1.0``, ``viscosity_inf = 0``, ``lambda=10`` and ``a=2`` ;
* (Second ``viscosity`` iteration) ``n=0.5``, ``viscosity_0 = 0.531702``, ``viscosity_inf = 0``, ``lambda=10`` and ``a=2`` 

and the first simulation uses the parameters in the **Physical Properties** section. For more information on ramping initial conditions, see :doc:`../../../parameters/cfd/initial_conditions`.

Particle
~~~~~~~~~~~~~~~~~~~~

In this case, we want to define a spherical boundary of radius 0.5 center at (0,0,0) that has no velocity. For more information on particle immersed boundary conditions using a sharp interface, see :doc:`../../../parameters/resolved_cfd-dem/resolved_cfd-dem`.

.. code-block:: text

	subsection particles
	  set number of particles = 1
	  set stencil order = 2
	  set length ratio  = 1
	  set refine mesh inside radius factor = 0.85
	  set refine mesh outside radius factor = 1.3
	  set initial refinement = 2
	  set integrate motion = false
	  set assemble Navier-Stokes inside particles = false    
	  subsection particle info 0

> 	  subsection position
> 		set Function expression =0;0;0
> 	  end
> 	  subsection velocity
> 		set Function expression =0;0;0
> 	  end
> 		subsection omega
> 		set Function expression =0;0;0
> 	  end

Reviewer 3: ```suggestion
	  subsection position
		set Function expression =0;0;0
	  end
```

		set pressure x =0.00001
		set pressure y =0.00001
		set pressure z =0.00001
		set radius = 0.5
	  end
	end

The hypershell around the boundary between ``refine mesh inside radius factor`` (r = 0.425) and ``refine mesh outside radius factor`` (r = 0.65) will initialy be refined twice (`initial refinement = 2`). 

Simulation control
~~~~~~~~~~~~~~~~~~~~~~~~~~

> The simulation is solved at steady-state with 2 mesh adaptation.

Reviewer 3: ```suggestion
The simulation is solved at steady-state with 2 mesh adaptations.
```

.. code-block:: text

	subsection simulation control
  	  set method                  = steady
	  set number mesh adapt       = 2
	  set output name             = sharp-carreau-output
	  set output frequency        = 1
	  set subdivision             = 1
	end

Mesh Adaptation Control
~~~~~~~~~~~~~~~~~~~~~~~~~~

> In order to generate the hypershell refinement zone around the immersed boundary, the ``mesh adaptation`` ``type`` must be set to ``kelly``. During both of the mesh refinement steps, 40% of the cells with be split in 8 (``fraction refinement   = 0.4``) using a velocity-gradient kelly operator.

Reviewer 3: ```suggestion
In order to generate an additional refinement zone around the immersed boundary, the ``mesh adaptation`` ``type`` must be set to ``kelly``. During both of the mesh refinement steps, 40% of the cells with be split in 8 (``fraction refinement   = 0.4``) using a velocity-gradient kelly operator.
```

.. code-block:: text

	subsection mesh adaptation
	  set type                  = kelly
	  set fraction coarsening   = 0.0
	  set fraction refinement   = 0.4

> 	  set fraction type	      	= number

Reviewer 3: ```suggestion
	  set fraction type	   = number
```

	  set frequency             = 1
	  set max number elements   = 8000000
	  set min refinement level  = 0
	  set max refinement level  = 11

> 	  set variable		      	= velocity

Reviewer 3: ```suggestion
	  set variable		     = velocity
```

	end

Results
---------------

Using Paraview, the steady-state velocity profile and the pressure profile can be visualized by operating a *slice* along the xy-plane (z-normal) that cuts in the middle of the sphere (See `documentation <https://forgeanalytics.io/blog/creating-slices-in-paraview/>`_). 

.. image:: images/velocity.png
	:align: center

.. image:: images/pressure.png
	:align: center

We can also see the viscosity profile throughout the domain, that is a function of the shear rate magnitude profile. Close to the particle, the shear rate is high which decreases the viscosity. 

.. image:: images/viscosity.png
	:align: center

.. image:: images/shear_rate.png
	:align: center

We can notice that the viscosity rapidely reaches a plateau at :math:`\eta=0.063`. Given the parameters in the **Physical properties** section, the viscosity behavior should be given  by:

.. image:: images/carreau.png
	:align: center

> We get the following force applied on the particle for each of the mesh refinements.

Reviewer 3: ```suggestion
We get the following torques and forces applied on the particle for each of the mesh refinements.
```

.. code-block:: text

    particle_ID    T_x       T_y       T_z      f_x      f_y       f_z    
	0 	-0.000008  0.000019 -0.000021 0.411248 0.000019  0.000038 
	0 	 0.000000  0.000001 -0.000003 0.415503 0.000001  0.000002 
	0 	-0.000000 -0.000000 -0.000000 0.424717 0.000002 -0.000002 
          
.. note:: 

> 	Since analysing non-Newtonian flow, there is no known solution for the drag coefficient. For a Newtonian flow at Re = 50, the drag force would be 0.6165. Therefore, the drag force was decreased usgin a shear-thinning fluid.

Reviewer 3: ```suggestion
	Because this analysis concerns non-Newtonian flow, there is no known solution for the drag coefficient. For a Newtonian flow at Re = 50, the drag force would be 0.6165. Therefore, the drag force was decreased using a shear-thinning fluid, as expected.
```

.. code-block:: text

    particle_ID    T_x       T_y       T_z      f_x      f_y       f_z    
	0 	-0.000008  0.000019 -0.000021 0.411248 0.000019  0.000038 
	0 	 0.000000  0.000001 -0.000003 0.415503 0.000001  0.000002 
	0 	-0.000000 -0.000000 -0.000000 0.424717 0.000002 -0.000002 
          
.. note:: 
	Because this analysis concerns non-Newtonian flow, there is no known solution for the drag coefficient. For a Newtonian flow at :math:`Re = 50`, the drag force would be :math:`0.6165`. Therefore, the drag force was decreased using a shear-thinning fluid.

Possibilities for extension
-----------------------------	
* **High-order methods** : Lethe supports higher order interpolation. This can yield much better results with an equal number of degrees of freedom than traditional second-order (Q1-Q1) methods, especially at higher Reynolds numbers.
* **Reynolds number** : By changing the inlet velocity, it can be interesting to see the impact of the shear-thinning behavior on the effective drag force.
* **Non-Newtonian parameters** : It can also be interesting to change the Carreau model parameters, i.e. changing the slope to appreciate the change in behavior.

> .. note::
> 	It is note possible to use the Carreau model for a shear-thickening flow. You should change the model to power-law, and use the decribed parameters in :doc:`../../../parameters/cfd/physical_properties`

Reviewer 3: ```suggestion
.. note::
	It is not possible to use the Carreau model for a shear-thickening flow. You should change the model to power-law, and use the decribed parameters in :doc:`../../../parameters/cfd/physical_properties`
```
